warehouse/oidc/views.py

```diff
@@ -0,0 +1,118 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import time
+
+from pyramid.view import view_config
+
+from warehouse.admin.flags import AdminFlagValue
+from warehouse.events.tags import EventTag
+from warehouse.macaroons import caveats
+from warehouse.macaroons.interfaces import IMacaroonService
+from warehouse.oidc.interfaces import IOIDCProviderService
+
+
+@view_config(
+    route_name="oidc.mint_token",
+    require_methods=["POST"],
+    renderer="json",
+    require_csrf=False,
+    has_translations=False,
+)
+def mint_token_from_oidc(request):
+    def _invalid(errors):
+        request.response.status = 422
+        return {"message": "Token request failed", "errors": errors}
+
+    oidc_enabled = request.registry.settings[
+        "warehouse.oidc.enabled"
+    ] and not request.flags.enabled(AdminFlagValue.DISALLOW_OIDC)
+    if not oidc_enabled:
+        return _invalid(
+            errors=[
+                {
+                    "code": "not-enabled",
+                    "description": "OIDC functionality not enabled",
+                }
+            ]
+        )
+
+    try:
+        body = request.json_body
+    except ValueError:
+        return _invalid(
+            errors=[{"code": "invalid-json", "description": "missing JSON body"}]
+        )
+
+    # `json_body` can return any valid top-level JSON type, so we have
+    # to make sure we're actually receiving a dictionary.
+    if not isinstance(body, dict):
+        return _invalid(
+            errors=[
+                {
+                    "code": "invalid-payload",
+                    "description": "payload is not a JSON dictionary",
+                }
+            ]
+        )
+
+    unverified_jwt = body.get("token")
+    if unverified_jwt is None:
+        return _invalid(
+            errors=[{"code": "invalid-token", "description": "token is missing"}]
+        )
+
+    if not isinstance(unverified_jwt, str):
+        return _invalid(
+            errors=[{"code": "invalid-token", "description": "token is not a string"}]
+        )
+
+    # For the time being, GitHub is our only OIDC provider.
+    # In the future, this should locate the correct service based on an
+    # identifier in the request body.
+    oidc_service = request.find_service(IOIDCProviderService, name="github")
+    provider, _claims = oidc_service.find_provider(unverified_jwt)
```

Comment:
I don't _love_ this interface, partly because of the multiple return values, and partly because this is still called `find_provider`.

Would it make sense to have two functions here, one being `verify_jwt` which takes an unverified JWT and returns the signed payload, if any, and a second `find_provider` which takes a verified JWT and returns the provider?

Reply:
Yeah, I agree. 

Part of why I joined them is because I want this API to be as misuse-resistant as possible: it should be impossible to look up an `OIDCProvider` without having a valid JWT corresponding to that provider, and it should be impossible to "verify" a JWT in a way that doesn't imply its intended use.

I think `verify_jwt` and `find_provider` would work, assuming signatures like this:

```python
def verify_jwt(unverified_token: str) -> VerifiedClaims: ...
def find_provider(verified_claims: VerifiedClaims) -> OIDCProvider: ...
```

...where `VerifiedClaims` is a `NewType` over the claim dictionary. That won't *prevent* API misuse, but it should help MyPy detect it. Thoughts?

Reply:
That seems acceptable to me.

Reply:
Great! I'll make those changes now.

Reply:
Oh, one tweak: `verify_jwt` should probably be `verify_jwt_signature`, since the "verification" we do in the context of returning the signed claims is *only* that the JWT has a valid and timely signature.

Reply:
[f48da18](https://github.com/pypi/warehouse/pull/11272/commits/f48da1898e31fc88240447e56f328a28a908828e) has the changes. I'll update the tests with a follow-up and confirm that everything still works.